multishard_design.md

```diff
@@ -0,0 +1,106 @@
+    Etcd [shape=box]
+    Root -> Etcd
+    Leaf1 -> Etcd
+    Leaf2 -> Etcd
+    Leaf3 -> Etcd
+  }
+)
+
+## Specifications
+### Sharding strategy
+We plan to use Document based sharding. A comparison between doc based sharding and term based sharding([stolen from Jeff Dean](http://web.stanford.edu/class/cs276/Jeff-Dean-Stanford-CS276-April-2015.pdf)):
+
+By doc: each shard has index for subset of docs 
+
+* pro: each shard can process queries independently
+* pro: easy to keep additional per-doc information 
+* pro: network traffic (requests/responses) small 
+* con: query has to be processed by each shard 
+* con: O(K*N) disk seeks for K word query on N shards Ways of Index Partitioning  
+
+By term: shard has subset of terms for all docs 
+
+* pro: K word query => handled by at most K shards
+* pro: O(K) disk seeks for K term query 
+* con: much higher network bandwidth needed. data about each term for each matching doc must be collected in one place
+* con: harder to have per-doc information
+
+### Service Discovery
+
+Our service discovery is simple: Root needs to know which leaf is up running and needs to be notified when one leaf dies away.  Etcd is used for service discovery. Once a leaf node is up, it registers itself into etcd and have a TTL attached to registered key-value pair. And leaf node will also start a goroutine to update the key-value pair periodically. Root node watches all registered key-value(actually a directory is being watched) and will be notified if a new leaf node is registered or a leaf node fails to update its key within TTL.
+
+### Metadata
+Metadata would be stored in Etcd. I am thinking currently we only need to store the mapping from shard to node. For more information about what is shard in our context, see case study of elasticsearch.
+
+### Persistent Storage
+Persistent storage is used by leaf node to checkpoint inverted index(or forward index as well). Ideally the storage should be something like a shared file system. Amazon s3 is a candidate.
+
+## Case Study
+### Elastic Search(version 2.x current)
+[Elasticsearch: The Definitive Guide](https://www.elastic.co/guide/en/elasticsearch/guide/current/index.html)
+
+ElasticSearch is a distributed search engine based on Lucene. When talking about multiple shards, in ElasticSearch, a shard is a Lucene index. In our case, we can treat a shard as a collections of inverted index.
+##### Route doc to shard
+`shard = hash(DocId) % num_of_shards`
+
+Questions: It is just a simple hash&mod function, how does Elasticsearch handles node failure? If a node fails, does it mean every doc has to recompute its shard and re-assign to a different shard?  
+No, Elasticsearch has a fixed number of shards. Even one node fails, the number of shards keeps unchanged. **One node can be assigned with multipe shards.** In case of node failure, elsticsearch will copy the entire shard from failed node to other node.(Does it mean there is a shared filesystem here?)
```

Comment:
Interesting! I understand why you said you think etcd only need to map shard to nodes.